Updated amp-accordion and amp-bind documentation (#26580)
ampproject/amphtml: extensions/amp-bind/amp-bind.md

  </tr>
</table>

## Usage

The `amp-bind` component enables custom stateful interactivity on AMP pages.

> For performance and to avoid the risk of unexpected content jumping, `amp-bind` does not evaluate expressions on page load. This means that the visual elements should be given a default state and not rely `amp-bind` for initial render.

Reviewer: suggestion: turn this into an `Important:` call out

Author: I have a theory that user's skim over the callouts. I think things that are this important should just be explained. I don't think reference docs should have a ton of callouts....

https://www.nngroup.com/articles/banner-blindness-old-and-new-findings/

> For performance and to avoid the risk of unexpected content jumping, `amp-bind` does not evaluate expressions on page load. This means that the visual elements should be given a default state and not rely on `amp-bind` for initial render.

Reviewer 2: ```suggestion
For performance, and to avoid the risk of unexpected content jumping, `amp-bind` does not evaluate expressions on page load. This means visual elements should be given a default state and not rely on `amp-bind` for initial render.
```

Glad to see this prominently featured.

    width="480" height="270"></amp-youtube>
<figcaption>Watch this video for an introduction to amp-bind.</figcaption></figure>

`amp-bind` has three main concepts:

> 1. [State](#state): A document-scope, mutable JSON state. State variables update in response to user actions. `amp-bind` does not evaluate expressions on page load. Visual elements should have their default "state" defined and not rely `amp-bind` for initial render.

Reviewer: s/not rely/not rely on

2. [Expressions](#expressions): JavaScript-like expressions that can reference the **state**.

> 3. [Bindings](#bindings): Special attributes that link an element's property to an **expression**. A property is bound by wrapping it inside brackets, in the form of `[property]`.

Reviewer: nit: the attribute is linked to **state** via an **expression**

### Basic example

[example preview="inline" playground="true" imports="amp-bind"]

```html
<p [text]="'Hello ' + foo">Hello World</p>

<button on="tap:AMP.setState({foo: 'Interactivity'})">
  Say "Hello Interactivity"
</button>
```

[/example]

In the example above:

- The **state** begins as empty.
- It has a single **binding** to `[text]`, the text content of a node, on the `<p>` element.
- The `[text]` value contains the **expression**, `'Hello ' + foo`.

>   - This expression concatenates the string 'Hello ' and the value of the **state variable** foo.

Reviewer: I wouldn't use a single item nested list

> - When the user taps/clicks the button, it triggers the `tap` event.

Reviewer: Suggestion: turn this into an enumerated list

```
When the user taps/clicks the button:

1. It triggers the `tap` event.
2. The `tap` event invokes the `AMP.setState()` method.
3. The `AMP.setState()` methods sets the `foo` **state variable** to the value of `Interactivity`.
4. The state is no longer empty, so the pages updates the bound property to its state.
```

Reviewer: Great explanation btw!

Author: ty, ty!

- The `tap` event invokes the `AMP.setState()` method.
- The `AMP.setState()` methods sets the `foo` **state variable** to the value of `Interactivity`.

> - The sate is no longer empty, so the pages updates the bound property to its state.

Reviewer: s/sate/state

Reviewer: s/pages/page

[tip type="note"]
Calling `AMP.setState()` in some examples may set or change states of other examples on page. Refresh this page to see examples before `AMP.setState()`.
[/tip]

### A slightly more complex example

[filter formats="websites, stories, ads"]

[example preview="top-frame" playground="true" imports="amp-bind"]

> ```html

Reviewer: missing `<head>`

  <style amp-custom>
    .greenBorder {
      border: 5px solid green;
    }
    .redBorder {
      border: 5px solid red;
    }
  </style>
</head>
<body>
<amp-state id="theFood">
  <script type="application/json">
    {
      "cupcakes": {
        "imageUrl": "https://amp.dev/static/samples/img/image2.jpg",
        "style": "greenBorder"
      },
      "sushi": {
        "imageUrl": "https://amp.dev/static/samples/img/image3.jpg",
        "style": "redBorder"
      }
    }
  </script>
</amp-state>
<div class="greenBorder" [class]="theFood[currentMeal].style">
  <p>Each food has a different border color.</p>
  <p [text]="'I want to eat ' + currentMeal + '.'">I want to eat cupcakes.</p>
  <amp-img
  width="300"
  height="200"
  src="https://amp.dev/static/samples/img/image2.jpg"
  [src]="theFood[currentMeal].imageUrl">
  </amp-img>
  <button on="tap:AMP.setState({currentMeal: 'sushi'})">Set to sushi</button>
  <button on="tap:AMP.setState({currentMeal: 'cupcakes'})">Set to cupcakes</button>
</div>
```

[/example]

In the example above:

> - The `<amp-state>` component holds the complex nested JSON data. It has an `id` of "theFood" to allow us to reference the defined data. But because `<amp-bind>` does not evaluate `<amp-state>` on page load, the **state** is empty.

Reviewer: s/"theFood"/`theFood`

> - The `<amp-state>` component holds the complex nested JSON data. It has an `id` of "theFood" to allow us to reference the defined data. But because `<amp-bind>` does not evaluate `<amp-state>` on page load, the **state** is empty.

Reviewer: `the complex nested JSON data` sounds a bit weird, maybe: `declares state using a JSON string literal.`

- The page loads with visual defaults.
  - The `<div>` element has `class="greenBorder"` defined.
  - The second `<p>` element has "I want cupcakes." defined within the tags.
  - The `<amp-img>` `src` points to a url.
- Changeable elements have **bindings** that point to **expressions**.
  - The `[class]` attribute on the `<div>` is bound to the `theFood[currentMeal].style` **expression**.
  - The `[text]` attribute on the second `<p>` is bound to the `'I want to eat ' + currentMeal + '.'` **expression**.
  - The `[src]` attribute is bound to the `theFood[currentMeal].imageUrl` **expression**.

If a user clicks the "Set to sushi" button:

> - The `tap` event trigger the `AMP.setState` action.

Reviewer: use enumaratin

- `style` is the **state variable**. It corresponds to the matching JSON key, and sets the bound property to that key's value.

[/filter] <!-- formats="websites, stories, ads" -->

[filter formats="email"]

[example preview="top-frame" playground="true" imports="amp-bind"]

```html
  <style amp-custom>
    .greenBorder {
      border: 5px solid green;
    }
    .redBorder {
      border: 5px solid red;
    }
  </style>
</head>
<body>
<amp-state id="theFood">
  <script type="application/json">
    {
      "cupcakes": {
        "style": "greenBorder",
        "text": "Just kidding, I want to eat cupcakes."
      },
      "sushi": {
        "style": "redBorder",
        "text": "Actually, I want to eat sushi."
      }
    }
  </script>
</amp-state>
<div class="greenBorder" [class]="theFood[currentMeal].style">
  <p>Each food has a different border color.</p>
  <p [text]="theFood[currentMeal].text">I want to eat cupcakes.</p>
  <button on="tap:AMP.setState({currentMeal: 'sushi'})">Set to sushi</button>
  <button on="tap:AMP.setState({currentMeal: 'cupcakes'})">Set to cupcakes</button>
</div>
```

[/example]

> - The `<amp-state>` component holds the complex nested JSON data. It has an `id` of "theFood" to allow us to reference the defined data. But because `<amp-bind>` does not evaluate `<amp-state>` on email load, the **state** is empty.

Reviewer: same as above

- `currentMeal` is the state name. In the case of `theFood` it will be `cupcakes` or `sushi`.
- `style` is the **state variable**. It corresponds to the matching JSON key, and sets the bound property to that key's value.

[/filter] <!-- formats="email" -->

### `<amp-state>` specification

[filter formats="websites, stories, ads"]
An `amp-state` element may contain either a child `<script>` element **OR** a `src` attribute containing a CORS URL to a remote JSON endpoint, but not both.

```html
<amp-state id="myLocalState">
  <script type="application/json">
    {
      "foo": "bar"
    }
  </script>
</amp-state>

<amp-state id="myRemoteState" src="https://data.com/articles.json"> </amp-state>
```

[/filter] <!-- formats="websites, stories, ads" -->

[filter formats="email"]
An `amp-state` element must contain a child `<script>` element.

```html
<amp-state id="myLocalState">
  <script type="application/json">
    {
      "foo": "bar"
    }
  </script>
</amp-state>
```

[/filter] <!-- formats="email" -->

[filter formats="websites, stories, ads"]

#### Attributes

[filter formats="websites, stories, ads"]

##### `src` (optional)

> The URL of the remote endpoint that will return the JSON that will update this `amp-state`. This must be a CORS HTTP service. The `src` attribute allows all standard URL variable substitutions. See the [Substitutions Guide](../../spec/amp-var-substitutions.md) for more info.

Reviewer: duplication: ... that will ...that will...

                "e": "nested five times",
                "ef": {
                  "f": "nested six times",
                  "fg": {
                    "g": "nested seven times",
                    "gh": {
                      "h": "nested nine times",
                      "hi": {
                        "i": "nested ten times"
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  </script>
</amp-state>
<p [text]="myState.foo"></p>
<p [text]="myState.first.ab.bc.cd.de.ef.fg.gh.hi.i"></p>
<button on="tap:AMP.setState({})">See value of set state</button>
<button
  on="tap:AMP.setState({ myState: {first: {ab: {bc: {cd: {de: {ef: {fg: {gh: {hi: {i:'this is as far as you should merge nested values'} } } } } } } } } } })"
>
  Merge 10th nested object
</button>
```

[/example]

#### Circular references

`AMP.setState(object)` throws an error if `object` contains a circular reference.

#### Removing a variable

Remove an existing state variable by setting its value to `null` in `AMP.setState()`.

```html
<button on="tap:AMP.setState({removeMe: null})"></button>
```

> #### Deep-merge with `AMP.setState()`

Reviewer: Hasn't this been explained before?

Author: Sort of. It doesn't go into detail about re-writing the JSON object...

Calling `AMP.setState()` deep-merges the provided object literal with the current state. `amp-bind` writes all literals to the state directly, except for nested objects, which are recursively merged. Primitives and arrays are in the state are always overwritten by variables of the same name in the object literal.

[example preview="inline" playground="true" imports="amp-bind"]

```html
<p [text]="employee.name">Name</p>
<p [text]="employee.age">Age</p>
<p [text]="employee.vehicle">Vehicle</p>
<!-- Pressing this button changes state to: -->
<button
  on="tap:AMP.setState({
              employee: {
                name: 'John Smith',
                age: 47,
                vehicle: 'Car'
              }
            })"
>
  Set employee to John Smith
</button>
<!-- Pressing this button recursively merges the object literal argument,  -->
<!-- `{employee: {age: 64}}`, into the existing state. -->

<button
  on="tap:AMP.setState({
              employee: {
                age: 64
              }
            })"
>
  Set employee age to 64
</button>
<!-- The value updates from 47 to 64 at employee.age.  -->
<!-- No other values change. -->
```

[/example]

### Refreshing state data

> The `refresh` action refetches data from data point the `src` attribute points to. It ignores the browser cache.

Reviewer: Shouldn't this be in an `actions` section?

[example preview="inline" playground="true" imports="amp-bind"]

```html
<amp-state id="myObjectState">
  <script type="application/json">
    {
      "hello": "world",
      "foo": "bar"
    }
  </script>
</amp-state>
<p [text]="'keys: ' + keys(myObjectState)">
  keys: myObjectState JSON object keys
</p>
<p [text]="'values: ' + values(myObjectState)">
  values: myObjectState JSON object values
</p>
<button on="tap:AMP.setState({})">Evaluate</button>
```

[/example]

#### [`Global`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects)

- [encodeURI](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURI)
- [encodeURIComponent](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent)

[example preview="inline" playground="true" imports="amp-bind"]

```html
<p [text]="'encodeURI: ' + encodeURI('https://amp.dev/😉')">
  encodeURI: Encode a URI and ignore protocol prefix
</p>
<p [text]="'encodeURIComponent: ' + encodeURIComponent('https://amp.dev/😉')">
  encodeURIComponent: Encode a URI
</p>
<button on="tap:AMP.setState({})">Evaluate</button>
```

[/example]

> #### Expression grammar

Reviewer: move this to the bottom

When the **state** changes, expressions tied to that state are evaluated. The element properties **bound** to the **state** are updated with the new expression results.

Boolean expression results toggle boolean attributes. For example: `<amp-video [controls]="expr"...>`. When `expr` evaluates to `true`, the `<amp-video>` element has the `controls` attribute. When `expr` evaluates to `false`, the `controls` attribute is removed.

[example preview="inline" playground="true" imports="amp-bind, amp-video"]

```html
<amp-video
  [controls]="controls"
  width="640"
  height="360"
  layout="responsive"
  poster="/static/inline-examples/images/kitten-playing.png"
>
  <source
    src="/static/inline-examples/videos/kitten-playing.webm"
    type="video/webm"
  />
  <source
    src="/static/inline-examples/videos/kitten-playing.mp4"
    type="video/mp4"
  />
  <div fallback>
    <p>This browser does not support the video element.</p>
  </div>
</amp-video>
<button on="tap:AMP.setState({ controls: true })">
  Controls
</button>
<button on="tap:AMP.setState({ controls: false })">
  No Controls
</button>
```

[/example]

> ### XML compatibility

Reviewer: suggestion: React and XML compatibility

Bind `Node.textContent` using the `[text]` attribute. The `[text]` attribute is supported on most text elements.

```html
<p [text]="'Hello ' + myState.foo">Hello World</p>
<p></p>
```

**CSS classes**

Bind an element's `class` using the `[class]` attribute. A `[class]` expression must result in a space-delimited string. Meaning, if you are binding multiple classes, use a space between names. A comma or dash will be evaluated as the class name.

[example preview="inline" playground="true" imports="amp-bind"]

```html
<style amp-custom>
  .background-green {
    background: green;
  }
  .background-red {
    background: red;
  }
  .border-red {
    border-color: red;
    border-width: 5px;
    border-style: solid;
  }
</style>
<div class="background-red" [class]="myClass">Hello World</div>
<!-- This button adds both classes -->
<button on="tap:AMP.setState({ myClass: 'background-green border-red' })">
  Working: Change Class
</button>
<!-- This expression evaluates to class="background-green,border-red" -->
<button on="tap:AMP.setState({ myClass: 'background-green,border-red' })">
  Broken: Change Class
</button>

> ```

Reviewer: mention that you can also provide a string array

Reviewer: `myClass: ['background-green'] ,['border-red']`

    width="480" height="270"></amp-youtube>
<figcaption>Watch this video for an introduction to amp-bind.</figcaption></figure>

`amp-bind` has three main concepts:

1. [State](#state): A document-scope, mutable JSON state. State variables update in response to user actions. `amp-bind` does not evaluate expressions on page load. Visual elements should have their default "state" defined and not rely `amp-bind` for initial render.
2. [Expressions](#expressions): JavaScript-like expressions that can reference the **state**.
3. [Bindings](#bindings): Special attributes that link an element's property to a **state** via an **expression**. A property is bound by wrapping it inside brackets, in the form of `[property]`.

> ### Basic example

Reviewer 2: Could we omit 'basic'?

[example preview="inline" playground="true" imports="amp-bind"]

```html
<p [text]="'Hello ' + foo">Hello World</p>

<button on="tap:AMP.setState({foo: 'Interactivity'})">
  Say "Hello Interactivity"
</button>
```

[/example]

In the example above:

- The **state** begins as empty.
- It has a single **binding** to `[text]`, the text content of a node, on the `<p>` element.

> - The `[text]` value contains the **expression**, `'Hello ' + foo`. This expression concatenates the string 'Hello ' and the value of the **state variable** foo.

Reviewer 2: Like this change too, 'string literal' while accurate is overboard for the example.

When the user taps/clicks the button:

1. It triggers the `tap` event.

> 1. The `tap` event invokes the `AMP.setState()` method.

Reviewer 2: Are these ok to be all the number 1?

Author: Markdown converts them into their proper numbers :)

1. The `AMP.setState()` methods sets the `foo` **state variable** to the value of `Interactivity`.
1. The state is no longer empty, so the page updates the bound property to its state.

[tip type="note"]

> Calling `AMP.setState()` in some examples may set or change states of other examples on page. Refresh this page to see examples before `AMP.setState()`.

Reviewer 2: Possible for us to iframe each example so they don't share a single global state? Or use different values in each example?

Author: We're going to work through the examples in the next iteration. Its just too much right now. Thanks for the suggestions!

```html
<head>
  <style amp-custom>
    .greenBorder {
      border: 5px solid green;
    }
    .redBorder {
      border: 5px solid red;
    }
  </style>
</head>
<body>
  <amp-state id="theFood">
    <script type="application/json">
      {
        "cupcakes": {
          "imageUrl": "https://amp.dev/static/samples/img/image2.jpg",
          "style": "greenBorder"
        },
        "sushi": {
          "imageUrl": "https://amp.dev/static/samples/img/image3.jpg",
          "style": "redBorder"
        }
      }
    </script>
  </amp-state>
  <div class="greenBorder" [class]="theFood[currentMeal].style">
    <p>Each food has a different border color.</p>
    <p [text]="'I want to eat ' + currentMeal + '.'">I want to eat cupcakes.</p>
    <amp-img
      width="300"
      height="200"
      src="https://amp.dev/static/samples/img/image2.jpg"
      [src]="theFood[currentMeal].imageUrl"
    >
    </amp-img>
    <button on="tap:AMP.setState({currentMeal: 'sushi'})">Set to sushi</button>
    <button on="tap:AMP.setState({currentMeal: 'cupcakes'})">
      Set to cupcakes
    </button>
  </div>
</body>
```

[/example]

In the example above:

> - The `<amp-state>` component declares state using a JSON string literal. It has an `id` of `theFood` to allow us to reference the defined data. But because `<amp-bind>` does not evaluate `<amp-state>` on page load, the **state** is empty.

Reviewer 2: Can likely drop 'string literal' or 'JSON' here.

```suggestion
- The `<amp-state>` component declares state using JSON. The `<amp-state>` element has an `id` of `theFood` to allow us to reference the defined data. But because `<amp-bind>` does not evaluate `<amp-state>` on page load, the **state** is empty.
```

- The page loads with visual defaults.
  - The `<div>` element has `class="greenBorder"` defined.
  - The second `<p>` element has "I want cupcakes." defined within the tags.
  - The `<amp-img>` `src` points to a url.
- Changeable elements have **bindings** that point to **expressions**.
  - The `[class]` attribute on the `<div>` is bound to the `theFood[currentMeal].style` **expression**.
  - The `[text]` attribute on the second `<p>` is bound to the `'I want to eat ' + currentMeal + '.'` **expression**.
  - The `[src]` attribute is bound to the `theFood[currentMeal].imageUrl` **expression**.

If a user clicks the "Set to sushi" button:

1. The `tap` event trigger the `AMP.setState` action.
1. The setState action turns `currentMeal` into a state and sets it to `sushi`.
1. AMP evaluates **bindings** with **expressions** that contain the state `currentMeal`.
1. `[class]="theFood[currentMeal].style"` updates `class` to `redBorder`.
1. `[text]="'I want to eat ' + currentMeal + '.'"` updates the inner text of the second `<p>` element to "I want to eat sushi".

> 1. `[src]="theFood[currentMeal].imageUrl` updates the `src` of `<amp-img>` to `https://amp.dev/static/samples/img/image3.jpg`

Reviewer 2: Here are these all #1 still or is this a shorthand for numerical items in a list in markdown?

Author: ;)

Author: Its so you don't mess up when you count and can move/remove/add items with ease. yay markdown!

- `currentMeal` is the state name. In the case of `theFood` it will be `cupcakes` or `sushi`.
- `style` is the **state variable**. It corresponds to the matching JSON key, and sets the bound property to that key's value.

[/filter] <!-- formats="email" -->

### `<amp-state>` specification

[filter formats="websites, stories, ads"]
An `amp-state` element may contain either a child `<script>` element **OR** a `src` attribute containing a CORS URL to a remote JSON endpoint, but not both.

```html
<amp-state id="myLocalState">
  <script type="application/json">
    {
      "foo": "bar"
    }
  </script>
</amp-state>

<amp-state id="myRemoteState" src="https://data.com/articles.json"> </amp-state>
```

[/filter] <!-- formats="websites, stories, ads" -->

[filter formats="email"]
An `amp-state` element must contain a child `<script>` element.

```html
<amp-state id="myLocalState">
  <script type="application/json">
    {
      "foo": "bar"
    }
  </script>
</amp-state>
```

[/filter] <!-- formats="email" -->

[filter formats="websites, stories, ads"]

#### Attributes

[filter formats="websites, stories, ads"]

##### `src` (optional)

> The URL of the remote endpoint that will return the JSON and update this `amp-state`. This must be a CORS HTTP service. The `src` attribute allows all standard URL variable substitutions. See the [Substitutions Guide](../../spec/amp-var-substitutions.md) for more info.

Reviewer 2: ```suggestion
The URL of the remote endpoint that must return JSON, which is used to this `amp-state`. This must be a HTTP service with a proper CORS configuration for the page. The `src` attribute allows all standard URL variable substitutions. See the [Substitutions Guide](../../spec/amp-var-substitutions.md) for more info.
```

Reviewer 2: Another nit here, but reinforcing these responses have to be JSON and correctly configured might help readers get on the path to success earlier.

[tip type="important"]
The endpoint must implement the requirements specified in the [CORS Requests in AMP](https://amp.dev/documentation/guides-and-tutorials/learn/amp-caches-and-cors/amp-cors-requests) spec.
[/tip]

##### `credentials` (optional)

Defines a `credentials` option as specified by the [Fetch API](https://fetch.spec.whatwg.org/).

- Supported values: `omit`, `include`
- Default: `omit`

To send credentials, pass the value of `include`. If this value is set, the response must follow the [AMP CORS security guidelines](https://amp.dev/documentation/guides-and-tutorials/learn/amp-caches-and-cors/amp-cors-requests/#cors-security-in-amp).
[/filter] <!-- formats="websites, stories, ads" -->

[filter formats="email"]

#### Actions

##### `refresh`

> The `refresh` action refetches data from data point the `src` attribute points to. It ignores the browser cache.

Reviewer 2: ```suggestion
The `refresh` action refetches data from data point the `src` attribute points to. This action will make a network request bypassing the browser's caching mechanisms.
```

Reviewer 2: Also in the future we might include how this request bypasses the network cache (I assume it's via `Math.random` query parameter).

[example preview="inline" playground="true" imports="amp-bind"]

```html
<amp-state id="currentTime" src="/documentation/examples/api/time"></amp-state>
<button on="tap:currentTime.refresh">
  Refresh
</button>
<div [text]="currentTime.time"></div>
```

[/example]

> A common need for `refresh` is working with live content that needs to be updated, such as a live sports score.

Reviewer 2: Might be worth a note that we think AMP Script is a better fit for live content, but `amp-bind` will work for a subset of cases.

                "ef": {
                  "f": "nested six times",
                  "fg": {
                    "g": "nested seven times",
                    "gh": {
                      "h": "nested nine times",
                      "hi": {
                        "i": "nested ten times"
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  </script>
</amp-state>
<p [text]="myState.foo"></p>
<p [text]="myState.first.ab.bc.cd.de.ef.fg.gh.hi.i"></p>
<button on="tap:AMP.setState({})">See value of set state</button>
<button
  on="tap:AMP.setState({ myState: {first: {ab: {bc: {cd: {de: {ef: {fg: {gh: {hi: {i:'this is as far as you should merge nested values'} } } } } } } } } } })"
>
  Merge 10th nested object
</button>
```

[/example]

#### Circular references

`AMP.setState(object)` throws an error if `object` contains a circular reference.

#### Removing a variable

Remove an existing state variable by setting its value to `null` in `AMP.setState()`.

```html
<button on="tap:AMP.setState({removeMe: null})"></button>
```

#### Deep-merge with `AMP.setState()`

> Calling `AMP.setState()` deep-merges the provided object literal with the current state. `amp-bind` writes all literals to the state directly, except for nested objects, which are recursively merged. Primitives and arrays are in the state are always overwritten by variables of the same name in the object literal.

Reviewer 2: This is a great explanation. Awesome!

Author: :)

[example preview="inline" playground="true" imports="amp-bind"]

```html
<amp-state id="count">
  <script type="application/json">
    1
  </script>
</amp-state>
<div>Item <span [text]="count">1</span></div>
<button on="tap:AMP.pushState({ count: count + 1 })">Increase count</button>
```

[/example]

Using `AMP.pushState()` sets the current state to the most recent pushed state.

## Expressions

`amp-bind` uses JavaScript-like expressions that can reference the state.

### Differences from JavaScript

- Expressions may only access the containing document's [state](#state).
- Expressions **do not** have access to `window` or `document`. `global` references the top-level state.
- Only `amp-bind` [allowed-listed functions](#allowed-listed functions) and operators are usable. are usable. Use of arrow functions are allowed as function parameters, e.g. `[1, 2, 3].map(x => x + 1)`.
  - Custom functions, classes and loops are disallowed.
- Undefined variables and array-index-out-of-bounds return `null` instead of `undefined` or throwing errors.

> - A single expression is currently capped at 50 operands for performance. Please [contact us](https://github.com/ampproject/amphtml/issues/new) if this is insufficient for your use case.

Reviewer 2: I thought this limit was recently upped. Might be worth coming back validating in a followup PR.

Author: Noted! Thank you.

- [concat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/concat)
- [filter](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter)
- [includes](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/includes)
- [indexOf](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/indexOf)
- [join](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/join)
- [lastIndexOf](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/lastIndexOf)
- [map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map)
- [reduce](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reduce)
- [slice](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice)
- [some](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some)

> - [sort](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort) (not-in-place)

Reviewer 2: This says `not-in-place` but there is an example for it below, should this omit `not-in-place`?

[example preview="inline" playground="true" imports="amp-bind"]

```html
<p [text]="'encodeURI: ' + encodeURI('https://amp.dev/😉')">
  encodeURI: Encode a URI and ignore protocol prefix
</p>
<p [text]="'encodeURIComponent: ' + encodeURIComponent('https://amp.dev/😉')">
  encodeURIComponent: Encode a URI
</p>
<button on="tap:AMP.setState({})">Evaluate</button>
```

[/example]

### Defining macros with `amp-bind-macro`

Reuse `amp-bind` expression fragments by defining an `amp-bind-macro`. The `amp-bind-macro` element allows an expression that takes zero or more arguments and references the current state. Invoke `amp-bind-macros` like a function, referencing the `id` attribute value from anywhere in the document.

[example preview="inline" playground="true" imports="amp-bind"]

```html
<amp-bind-macro
  id="circleArea"
  arguments="radius"
  expression="3.14 * radius * radius"
></amp-bind-macro>
<p>
  Input a radius value
</p>
<input
  type="number"
  min="0"
  max="100"
  value="0"
  on="input-throttled:AMP.setState({myCircle:{radius: event.value}})"
/>
<p>
  The circle has an area of
  <span [text]="circleArea(myCircle.radius)">0</span>.
</p>
```

[/example]

> A macro can also call other macros <i>defined before itself</i>. A macro cannot call itself recursively.

Reviewer 2: Is there a limit to the number of macros a macro can call? Worth documenting in a followup?

Author: also added to notes

### Binding types

`amp-bind` supports data bindings on five types of element state.

[**Node.textContent**](https://developer.mozilla.org/en-US/docs/Web/API/Node/textContent)

Bind `Node.textContent` using the `[text]` attribute. The `[text]` attribute is supported on most text elements.

```html
<p [text]="'Hello ' + myState.foo">Hello World</p>
<p></p>
```

**CSS classes**

Bind an element's `class` using the `[class]` attribute. A `[class]` expression must result in a space-delimited string. Meaning, if you are binding multiple classes, use a space between names. A comma or dash will be evaluated as the class name.

[example preview="inline" playground="true" imports="amp-bind"]

```html
<style amp-custom>
  .background-green {
    background: green;
  }
  .background-red {
    background: red;
  }
  .border-red {
    border-color: red;
    border-width: 5px;
    border-style: solid;
  }
</style>
<div class="background-red" [class]="myClass">Hello World</div>
<!-- This button adds both classes -->
<button on="tap:AMP.setState({ myClass: 'background-green border-red' })">
  Working: Change Class
</button>
<!-- String arrays also work -->
<button

>   on="tap:AMP.setState({ myClass: ['background-green'] ,['border-red'] })"

Reviewer 2: ```suggestion
  on="tap:AMP.setState({ myClass: ['background-green'], ['border-red'] })"
```

  Broken: Change Class
</button>
```

[/example]

[**the `hidden` attribute**](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/hidden)

Hide and reveal and element using the `[hidden]` attribute. A `[hidden]` expression should be a boolean expression.

[example preview="inline" playground="true" imports="amp-bind"]

```html
<p [hidden]="hiddenState">Hello there!</p>
<button on="tap:AMP.setState({hiddenState: true})">Hide</button>
<button on="tap:AMP.setState({hiddenState: false})">Show</button>
```

[/example]

**Size of [AMP components](https://www.ampproject.org/docs/reference/components)**

Change the `width` and `height` using the `[width]` and `[height]` attributes.

[example preview="inline" playground="true" imports="amp-bind"]

```html
<amp-img
  src="https://unsplash.it/400/200"
  width="200"
  [width]="myImageDimension.width"
  height="100"
  [height]="myImageDimension.height"
>
</amp-img>
<button
  on="tap:AMP.setState({
              myImageDimension: {
              width: 400,
              height: 200
              }
              })"
>
  Change size
</button>
```

[/example]

**Accessibility states and properties**

> Use to dynamically update information available to assistive technologies, such as screen readers. All `[aria-*]` and `[data-*]` are bindable. See the [full list here](https://www.w3.org/WAI/PF/aria-1.1/states_and_properties).

Reviewer 2: 👍

> ##### `src` (optional)

Reviewer: header too deep?

Author: Nope. Usually Attributes gets a top heading, but this is specific to the `<amp-state>` component, which is not needed to use `amp-bind`. I guess it's a technicality?

[filter formats="email"]

> #### Actions

Reviewer: header too deep?

Author: Same no reason as above...
